feat: add universal agent id standard

kantorcodes opened the pull request:
## Add HCS-14: Universal Agent ID Standard

The HCS-14 standard provides a systematic approach for generating globally unique identifiers for AI agents using the W3C Decentralized Identifier (DID) framework. By introducing dual DID methods (AID for registry-generated and UDID for self-sovereign identifiers), this standard enables consistent agent identification across web2 APIs, web3 protocols, and hybrid systems without requiring any specific infrastructure dependencies. Both methods embed routing information directly in the DID, making agent discovery and communication seamless across protocol boundaries.

### What it does
- Generates consistent IDs like `did:udid:QmX4fB9XpS3yKqP8MHTbcQW7R6wN4PrGHz;registry=hol;nativeId=302a300506032b6570032100e7d59d8bff3f9e1784cd4e7f340fb1a7333ee264fed4beb0b38fe7e4d29d04;uid=0.0.123456` 
- Uses optional flags that provide additional context around the ID for lookups. 

### Motivation 
Every protocol has its own ID system. HCS-14 creates a universal ID that works everywhere, enabling cross-protocol agent discovery and communication that is practical for both Web2 and Web3. 

Supports A2A, MCP, HCS-10, NANDA, OLAS, Virtuals, and more.

itsbrandondev commented:
had a chance to review, this looks really great. will be useful for quite a few things we are working on and i think is a big step forward for the industry. no feedback to provide other than voicing my support! great work.

kantorcodes commented on docs/standards/hcs-14.md:
```diff
@@ -0,0 +1,647 @@
+- `hash` = Base58-encoded identifier hash
+- `parameters` = Semicolon-separated key-value pairs for routing
+
+### DID Parameter Structure
```
- What learnings can we take from enterprises for the generation of parameters 
- Taxonomy systems are very important e.g. SNOMED, LOINC, ICD9 versus ICD10
- Attributes can include purpose 
     - Purpose can have sub-purposes (graph structure, top level and down)
     - did:{method}:{hash};purpose[0]=0;purpose[0][1]=2;purpose[1]=1 (consider renaming purpose to `p`)

- Ahmer Inam

jaycoolh replied:
Is the purpose syntax definitely part of the ID? It's not specified in parameter definitions list below.

Would it make more sense to have it as part of the Canonical Agent Data which generates the hash? Alongside the skills etc

kantorcodes commented on docs/standards/hcs-14.md:
```diff
@@ -0,0 +1,647 @@
+
+The HCS-14 standard provides a systematic approach for generating globally unique identifiers for AI agents using the W3C Decentralized Identifier (DID) framework. By introducing dual DID methods (AID for registry-generated and UDID for self-sovereign identifiers), this standard enables consistent agent identification across web2 APIs, web3 protocols, and hybrid systems without requiring any spec ...
+
+## Motivation
```
Traceability should be a core motivation as part of this standard

- Ahmer Inam

kantorcodes commented:
> had a chance to review, this looks really great. will be useful for quite a few things we are working on and i think is a big step forward for the industry. no feedback to provide other than voicing my support! great work.

Thanks, Brandon!

jaycoolh commented on docs/standards/hcs-14.md:
```diff
@@ -0,0 +1,648 @@
+
+AI agents are proliferating across diverse platforms with different identification systems. Each protocol has evolved sophisticated identification systems optimized for their specific use cases - Web2 platforms provide stable API-based identifiers, blockchain systems leverage cryptographic addresses, consensus-based networks utilize account IDs, and decentralized registries implement database-opt ...
+
+**Cross-Protocol Collaboration**: As agents become more sophisticated, they increasingly need to work together across protocol boundaries. A research agent on one protocol might need to collaborate with a data analysis agent on another, or a creative agent might need to coordinate with a transaction agent on a different network. Enabling seamless collaboration amplifies the capabilities of all ag ...
```
Are there any real life / more concrete examples of cross-protocol collaboration that could be given here?

jaycoolh commented on docs/standards/hcs-14.md:
```diff
@@ -0,0 +1,648 @@
+
+**Unified Discovery**: Each protocol has developed robust discovery mechanisms within their ecosystem. By creating a universal identification layer, we can build on these existing systems to enable agents to discover potential collaborators across all protocols, expanding the possibilities for agent interactions.
+
+**Trust Bridge Building**: Different protocols have implemented innovative trust and verification mechanisms suited to their architectures. A universal ID system can serve as a bridge between these trust models, allowing protocols to maintain their security approaches while enabling cross-protocol verification.
```
Does 'trust' here refer an Agent's reputation and track record at performing a task, or something else?

kantorcodes replied:
Great callout, this is addressed in "Trust and Reputation (Informative)" based on your feedback

jaycoolh commented on docs/standards/hcs-14.md:
```diff
@@ -0,0 +1,648 @@
+
+3. **Cryptographic Verifiability**: Any party can verify that an HCS-14 ID correctly represents an agent by regenerating it from the agent's public data. This creates a trustless system where the ID itself is the proof of authenticity, eliminating the need for certificate authorities or attestation services.
+
+4. **Collision-Resistant Design**: By using SHA-384 hashing with carefully structured canonical data, we achieve cryptographic assurance against ID collisions. The probability of two different agents generating the same ID is effectively zero (2^-384), ensuring global uniqueness without coordination.
```
SHA-384 is considered post-quantum too - maybe a nice to have detail to add in?

jaycoolh commented on docs/standards/hcs-14.md:
```diff
@@ -0,0 +1,648 @@
+    version: agentData.version.trim(),
+    protocol: agentData.protocol.toLowerCase().trim(),
+    nativeId: agentData.nativeId.trim(),
+    skills: (agentData.capabilities || []).sort((a, b) => a - b),
```
```suggestion
    skills: (agentData.skills || []).sort((a, b) => a - b),
```

to align with convention set above of naming `agentData` and `conical` object attributes the same

jaycoolh commented on docs/standards/hcs-14.md:
```diff
@@ -0,0 +1,648 @@
+    params.push(`uid=${agentData.uid}`);
+  }
+
+  return `did:udid:${base58Hash};${params.join(';')}`;
```
What happens if I create an agent with the same `agentData` as an already existing agent? and our did IDs are identical

jaycoolh commented on docs/standards/hcs-14.md:
```diff
@@ -0,0 +1,648 @@
+
+2. **Deterministic Serialization**: JSON objects MUST be serialized with keys sorted lexicographically. Skills arrays MUST be sorted numerically in ascending order.
+
+3. **Hash Generation**: SHA-256 MUST be applied to UTF-8 encoded canonical JSON. The resulting hash MUST be encoded using Base58.
```
SHA-256 or SHA-384 ?

```suggestion
3. **Hash Generation**: SHA-384 MUST be applied to UTF-8 encoded canonical JSON. The resulting hash MUST be encoded using Base58.
```

kantorcodes commented:
@codex review this PR

kantorcodes commented:
Follow-up: ERC-8004 + ToIP compatibility and ISO terminology

- Interop section: added an “Interoperability (Informative)” section with two subsections:
  - ERC-8004: UAID supports on-chain agents; recommend CAIP-10 `eip155:<chainId>:<address>` for `nativeId`; `registry` namespace can represent the ERC-8004 program (e.g., `olas`, `virtuals`). DID Documents may include verification methods derived from the EVM account and service endpoints for A2A/MCP.
  - Trust Over IP: clarified that did:uaid resolves to DID Core–conformant DID Documents with optional service entries; recommend `alsoKnownAs` links to protocol DIDs (did:pkh / did:ethr / did:web) for federation.
- Resolution profile: added a short “UAID DID Resolution Profile” defining minimal DID Document fields.
- Native IDs: updated EVM example to CAIP-10 in “Native Protocol IDs” and Example 5.
- Normative language: added a “Conformance and Normative Language” section and converted RFC-style MUST/SHOULD/MAY to ISO/IEC shall/should/may. Left an explicit note on interpretation for prior drafts.

All changes are local on `feat/hcs-14`; I can push whenever you’re ready.

kantorcodes commented:
Normative language sweep (ISO/IEC)

- Added a “Conformance and Normative Language” section referencing ISO/IEC Directives, Part 2 and an interpretation note for legacy RFC 2119/8174 terms (MUST/SHOULD/MAY → shall/should/may).
- Replaced remaining lowercase “must/must not” and uppercase forms with ISO terms where normative.
- Strengthened normalization bullets to explicit “shall” statements.
- Marked Hash Generation steps as normative (“implementations shall perform steps 1–6”).
- Clarified uid: “is required and shall be \"0\" if not applicable”.
- Left guidance as “should/may” where appropriate; informative sections are labeled as such.

Let me know if you prefer CAIP‑10 for EVM nativeId to be a “shall” (currently “should” in Interoperability + Native ID notes). Ready to push.

kantorcodes commented:
Added UAID usage examples

- A2A agent.json: New informative example shows embedding `did:uaid` in `/.well-known/agent.json` with routing params.
- HCS-11 profile: Added “UAID Integration (Informative)” section with a minimal snippet using `properties.uaid` and updated both example profiles to include a UAID.
- Cross-link: HCS-14 now includes a corresponding HCS-11 profile example.

Let me know if you prefer a top-level `did` field in HCS-11 instead of `properties.uaid`; I kept it non-breaking and informative for now.
